master: unify experiments enumeration (+ some tooling unit tests)

airwoodix opened the pull request:
<!--

Thank you for submitting a PR to ARTIQ!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

You can also read more about contributing to ARTIQ in this document:
https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#contributing-code

Based on https://raw.githubusercontent.com/PyCQA/pylint/master/.github/PULL_REQUEST_TEMPLATE.md
-->

# ARTIQ Pull Request

## Description of Changes

This patch:

  - refactors the enumeration of experiments from Python modules in ``artiq.master.worker_impl``. In particular, it now uses ``tools.get_experiment`` instead of an almost literal copy of it. This unifies the behavior with ``artiq_run`` and ``artiq_compile`` (or was there a reason not to exclude underscore-prefixed experiments in the master's ``get_exp``?);

  - rewrites ``tools.get_experiment`` with ``inspect.getmembers`` instead of ``__dict__``. Introduces a new predicate ``language.is_public_experiment``. The existing ``language.is_experiment`` is not used anywhere in the codebase but is part of the public API. These predicates (or maybe only one) should maybe be moved to ``artiq.tools`` (+ release notes entry);

  - rewrites ``master/worker_impl:examine`` with ``inspect.getmembers`` (simpler);

  - adds units tests for ``tools.file_import`` (I almost broke it in #1605)  and ``tools.get_experiment``.

## Type of Changes

<!-- Leave ONLY the corresponding lines for the applicable type of change: -->
|   | Type |
| ------------- | ------------- |
|   | :bug: Bug fix  |
|   | :sparkles: New feature |
| ✓  | :hammer: Refactoring  |
|   | :scroll: Docs |

## Steps (Choose relevant, delete irrelevant before submitting)

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Update [RELEASE_NOTES.rst](../RELEASE_NOTES.rst) if there are noteworthy changes, especially if there are changes to existing APIs.
- [x] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [x] Check, test, and update the [unittests in /artiq/test/](../artiq/test/) or [gateware simulations in /artiq/gateware/test](../artiq/gateware/test)


### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

sbourdeauducq commented on artiq/test/test_tools.py:
```diff
@@ -0,0 +1,125 @@
+
+
+def random_chars(length=8):
+    return "".join(random.choices(string.ascii_letters, k=length))
```
Probably not an issue here but it's generally good to avoid randomness in unit tests, to avoid non-reproducible results.

sbourdeauducq commented:
Thanks, looking great in general!

> This unifies the behavior with artiq_run and artiq_compile (or was there a reason not to exclude underscore-prefixed experiments in the master's get_exp?);

Excluding them makes it impossible to submit experiments via ``artiq_client`` that start with an underscore (which is why I had done it this way), though I suspect nobody ended up wanting to do that, and as you point out it is inconsistent with the behavior of other tools. Might still warrant a release notes entry.

airwoodix commented:
Thanks for the feedback. I removed the random module names and prefix in the unit tests and added an entry in the release notes.

> makes it impossible to submit experiments via artiq_client that start with an underscore

It still works with ``--class-name``.

I guess it's even a feature now: one can have multiple (dependent) experiments in a single module and still submit the entry point with a simple ``artiq_client submit`` command.
